corehq/apps/locations/util.py

```diff
@@ -179,55 +199,68 @@ def get_consumption(self, loc):
-            }
-            tab_rows.append(dict(flatten_json(loc_dict)))
-
-        header_keys = ['location_id', 'site_code', 'name', 'parent_code',
-                       'latitude', 'longitude', 'external_id', 'do_delete']
-        tab_headers = [LOCATION_SHEET_HEADERS[h] for h in header_keys]
-
-        def _extend_headers(prefix, headers):
-            tab_headers.extend(json_to_headers(
-                {prefix: {header: None for header in headers}}
-            ))
-        _extend_headers('data', (f.slug for f in self.data_model.fields))
-        _extend_headers('uncategorized_data', uncategorized_keys)
-        if self.include_consumption_flag and loc_type.name not in self.administrative_types:
-            _extend_headers('consumption', self.product_codes)
-
-        sheet_title = loc_type.code
-
-        return (sheet_title, {
-            'headers': tab_headers,
-            'rows': tab_rows,
-        })
+    def get_headers(self):
+        headers = OrderedDict([
+            ('types', [list(LOCATION_TYPE_SHEET_HEADERS.values())])
+        ])
+        for loc_type in self.location_types:
+            additional_headers = []
+            additional_headers.extend(self._prefix_headers('data', (f.slug for f in self.data_model.fields)))
+            if self.include_consumption_flag and loc_type.name not in self.administrative_types:
+                additional_headers.extend(self._prefix_headers('consumption', self.product_codes))
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['uncategorized_data'])
+            additional_headers.append(LOCATION_SHEET_HEADERS_OPTIONAL['delete_uncategorized_data'])
+
+            headers[loc_type.code] = [LOCATION_SHEET_HEADERS_BASE.values() + additional_headers]
+
+        return list(headers.items())
+
+    def write_data(self, writer):
+        self._write_type_sheet(writer)
+        for location_type in self.location_types:
+            self._write_locations(writer, location_type)
+
+    def _write_locations(self, writer, location_type):
+        include_consumption = self.include_consumption_flag and location_type.name not in self.administrative_types
+        query = SQLLocation.active_objects.filter(
+            domain=self.domain,
+            location_type__name=location_type.name
+        )
 
-    def type_sheet(self, location_types):
-        headers = LOCATION_TYPE_SHEET_HEADERS
+        def _row_generator(include_consumption=include_consumption):
+            for loc in query:
+                model_data, uncategorized_data = self.data_model.get_model_and_uncategorized(loc.metadata)
+                row = [
+                    loc.location_id,
+                    loc.site_code,
+                    loc.name,
+                    loc.parent.site_code if loc.parent else '',
+                    loc.latitude or '',
+                    loc.longitude or '',
+                    loc.external_id,
+                    '',  # do delete
+                ]
```

Comment:
nit: I prefer not to use positional orderings for stuff like this, but it's not a big deal.  My ideal is the API from [`csv.DictWriter`](https://docs.python.org/2/library/csv.html#csv.DictWriter).

Reply:
Agree but that's a bigger change than I want to make right now